rafthttp/peer.go

```diff
@@ -258,4 +260,43 @@ func (p *peer) pick(m raftpb.Message) (writec chan<- raftpb.Message, picked stri
 	return p.pipeline.msgc, pipelineMsg
 }
 
+type peerStatus struct {
+	id         types.ID
+	mu         sync.Mutex
+	active     bool
+	lastErrMap map[string]error
+}
+
+func newPeerStatus(id types.ID) *peerStatus {
+	return &peerStatus{
+		id:         id,
+		lastErrMap: make(map[string]error),
+	}
+}
+
+func (s *peerStatus) success() {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if !s.active {
+		log.Printf("the connection with %s became active", s.id)
+		s.active = true
+		s.lastErrMap = make(map[string]error)
+	}
+}
+
+func (s *peerStatus) failure(name, action string, err error) {
```

Comment:
i change the name to `deactivate` and use string for reason. I don't want to introduce failureType in thie level for the ease of caller usage.
I use failureType internally as map key for clarity.